xmtp/src/association.rs

```diff
@@ -0,0 +1,109 @@
+use crate::types::Address;
+use serde::{Deserialize, Serialize};
+use thiserror::Error;
+use xmtp_crypto::signature::{RecoverableSignature, SignatureError, SigningKey};
+use xmtp_crypto::utils::{self, eth_address};
+
+#[derive(Debug, Error)]
+pub enum AssociationError {
+    #[error("bad signature")]
+    BadSignature(#[from] SignatureError),
+    #[error("Association text mismatch")]
+    TextMismatch,
+    #[error(
+        "Address mismatch in Association: Provided:{provided_addr:?} != signed:{signing_addr:?}"
+    )]
+    AddressMismatch {
+        provided_addr: Address,
+        signing_addr: Address,
+    },
+    #[error("unknown association error")]
+    Unknown,
+}
+
+#[derive(Serialize, Deserialize, PartialEq, Debug)]
+pub enum AssociationText {
+    Static { addr: Address, key_bytes: Vec<u8> },
+}
+
+impl AssociationText {
+    pub fn get_address(&self) -> Address {
+        match self {
+            Self::Static { addr, .. } => addr.clone(),
+        }
+    }
+
+    pub fn text(&self) -> String {
+        match self {
+            Self::Static { addr, key_bytes } => gen_static_text_v1(addr, key_bytes),
+        }
+    }
+
+    pub fn is_valid(&self, addr: &str, key_bytes: &[u8]) -> Result<(), AssociationError> {
+        if self.text() == gen_static_text_v1(addr, key_bytes) {
```

Comment:
> is_valid() is almost like a reimplementation of text(), right?

is_valid is spiritually different from text() in the AssociationText context. 
- text() -> return the text that is held inside the association. For static text this can be currently be generated. Adding a timestamp would require that the text be saved.
- is_valid() -> is the AssociationText legally formatted. Does it conform to the rules.

While in the current Static case

Reply:
>Do we intend to validate a raw text string that we read directly from the network?

Contact bundles will contain an association and a public key. Clients will need to validate the AssociationText and signature to ensure the chain of trust between address and account has been established